### PR TITLE
Add a `--rules` flag to the script

### PR DESCRIPTION
**Pull request by srawlins**
This required a few significant refactorings:

* Each Rule now only knows its own name via a static String, rather than an instance String. ☹️. This has some negative consequences, but allows for String-based user APIs, like a config file, and the command line `--rules` arg.
* The engine now has a map from Rule names to Rule instances.
* Each reported Lint now only has only a reference to the `name` of the Rule that generated it, rather than a reference to the instance of the Rule. However, the only thing we've used so far, when reaching into a Lint's Rule, was its `name`.

### CODE REVIEW

**Comment by srawlins** [comment]
@nex3 I'm definitely open to suggestions on some of these changes; some of this feels kludgey to me.

File: lib/src/engine.dart [hunk 1]
```diff
@@ -55,11 +59,19 @@ class Engine {
...
       } else {
         var source = new File(path).readAsStringSync();
-        return new Linter(source, allRules, url: path).run();
+        return new Linter(source, rules, url: path).run();
       }
     }).expand((lint) => lint);
   }
 }
+
+Iterable<Rule> parseRules(List<String> rules) => rules.map((rule) {
```
[the remark this answers is not in the document]
Comment (srawlins) [reply lib/src/engine.dart]: Done.

File: lib/src/engine.dart [hunk 2]
```diff
@@ -17,13 +18,13 @@ import 'rules/use_falsey_null.dart';
...
 /// this list is used. That cannot happen until the binary supports a config
 /// file or a flag with a list of lint rules.
-final allRules = <Rule>[
-  new NoDebugRule(),
-  new NoEmptyStyleRule(),
-  new NoLoudCommentRule(),
-  new QuoteMapKeysRule(),
-  new UseFalseyNullRule(),
-];
+final ruleNamesMap = <String, Rule>{
```
[the remark this answers is not in the document]
Comment (srawlins) [reply lib/src/engine.dart]: Done.

File: lib/src/engine.dart [hunk 3]
```diff
@@ -34,10 +35,13 @@ class Engine {
   /// This list may also include "-", the special path that represents `stdin`.
   final List<String> paths;
 
+  final Iterable<Rule> rules;
+
   /// The file URL to use when reporting lints from stdin.
   final String stdinFileUrl;
 
-  Engine(this.paths, {this.stdinFileUrl});
+  Engine(this.paths, {Iterable<Rule> rules, this.stdinFileUrl})
```
[the remark this answers is not in the document]
Comment (nex3) [reply lib/src/engine.dart]: Probably also mention that the engine uses all available rules if `rules` isn't passed.

[the remark this answers is not in the document]
Comment (srawlins) [reply lib/src/engine.dart]: Done.

**Pull request merged**